Add basic implementation

jonnydgreen opened the pull request:
Closes: https://github.com/mercurius-js/auth/issues/1

PR so you can follow my progress as I head towards shipping v1.0 :)

Still todo:
 - [x] Add Application Lifecycle hook to Mercurius
 - [x] Make auth directive configurable
 - [x] Go over validation options and add more tests for these
 - [x] Define symbols for internals
 - [x] Use resolver overrides
 - [x] More tests for the gateway
 - [x] General tidy up of codebase
 - [x] Set custom auth errors
 - [x] Improve performance (using current Benchmarks with and without Auth as a reference)
 - [x] Docs
 - [x] Examples

mcollina commented on test/apply-policy.js:
```diff
@@ -0,0 +1,278 @@
+        identity: context.reply.request.headers['x-user']
+      }
+    },
+    applyPolicy: async (authDirectiveAST, context, field) => {
```
have we got access to the current object/paths that lead to the current resolver? I think we need to express a concept like "this object is accessible only from owners of the parent org".

jonnydgreen replied:
good point! We have partial access to the current object through the `field` argument with is a `GraphQLField` type, I'm also building a cursor object as I've traverse the AST, so can also expose that in a sensible manner. What I'll do is change this to closely match with what is available in a normal resolver, e.g. `schema`, `info`, `context`, field etc.

I'm also adding support for the user to provide custom errors as I reckon this'll be needed

mcollina commented on lib/auth.js:
```diff
@@ -0,0 +1,141 @@
+
+      const errors = await this[kApplyPolicyToSelections](schemaTypeMap, null, definition, context, { path: [] })
+      authErrors.push(...errors)
+    }))
```
I think this is not strictly correct as I think we need to apply these to the real objects instead. Maybe an approach that overrides the resolvers?

jonnydgreen replied:
Yeah that could work pretty nicely for a couple of reasons!
 - more performant - no need to traverse the query AST
 - easy access to all the parent objects etc straight away - solving: https://github.com/mercurius-js/auth/pull/3#discussion_r615383879

My only concern would be in gateway mode. I.e. when the federated schema changes, the resolvers wouldn't necessarily reflect the latest schema. I'm thinking of solving this by maybe adding a mercurius application lifecycle hook that triggers when the federated schema changes so we can re-override the resolvers?

mcollina replied:
agreed!

jonnydgreen replied:
Benchmarks are looking a lot better after refactoring to use resolver overrides! Still got a fair bit to polish up (including a PR in mercurius to make), but looking very promising!

```
==============================
= Normal Mode | Without Auth =
==============================
Running 10s test @ http://localhost:3000/graphql
100 connections
50 workers

┌─────────┬──────┬──────┬───────┬───────┬─────────┬────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev  │ Max    │
├─────────┼──────┼──────┼───────┼───────┼─────────┼────────┼────────┤
│ Latency │ 0 ms │ 8 ms │ 18 ms │ 23 ms │ 8.38 ms │ 4.5 ms │ 129 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%  │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 5391    │ 5391    │ 10679   │ 11023  │ 10137   │ 1624.79 │ 5391    │
├───────────┼─────────┼─────────┼─────────┼────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.84 MB │ 2.84 MB │ 5.62 MB │ 5.8 MB │ 5.33 MB │ 854 kB  │ 2.84 MB │
└───────────┴─────────┴─────────┴─────────┴────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

113k requests in 10.02s, 59.4 MB read
===========================
= Normal Mode | With Auth =
===========================
Running 10s test @ http://localhost:3000/graphql
100 connections
50 workers

┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max    │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼────────┤
│ Latency │ 0 ms │ 9 ms │ 19 ms │ 24 ms │ 9.16 ms │ 4.42 ms │ 171 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 5083    │ 5083    │ 9903    │ 10015   │ 9407.8  │ 1447.56 │ 5082    │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.67 MB │ 2.67 MB │ 5.21 MB │ 5.27 MB │ 4.95 MB │ 761 kB  │ 2.67 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

104k requests in 10.02s, 54.8 MB read
```

jonnydgreen commented:
Things are way simpler now and it runs ~40% faster after the override field refactor (thanks for the suggestion!) - I've started preparing a PR to add application lifecycle hooks to Mercurius and will get that submitted asap! Then, it's a case of getting the docs and examples together and we should be pretty much good to go!

mcollina commented on bench/gateway-bench.js:
```diff
@@ -0,0 +1,45 @@
+const instance = autocannon(
+  {
+    url: 'http://localhost:3000/graphql',
+    workers: 50,
```
drop this workers option, not needed here

mcollina replied:
workers was not removed.

jonnydgreen replied:
my apologies, it's removed locally - just not pushed as the auth doesn't fully work when JIT is enabled and I wanted to avoid a push where the tests are broken (I'll push now so you can see the code!). This is fixed in the Mercurius PR I submitted yesterday :) https://github.com/mercurius-js/mercurius/pull/461. In the future, I'll make sure to only resolve a comment after a push :)

mcollina commented on bench/normal-bench.js:
```diff
@@ -0,0 +1,44 @@
+const instance = autocannon(
+  {
+    url: 'http://localhost:3000/graphql',
+    workers: 50,
```
drop this workers option, not needed here

mcollina commented on bench/gateway-without-auth.js:
```diff
@@ -0,0 +1,20 @@
+const app = Fastify()
+
+app.register(mercurius, {
+  gateway: {
```
There is probably a bit more work to change the default settings, as I think they might be limiting the perf here.

jonnydgreen replied:
sure thing, I'll take a look at the mercurius options to see if I can improve perf

mcollina commented on lib/auth.js:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  async registerAuthHandlers (schema) {
```
can you drop async?

jonnydgreen replied:
good spot! Yep, I'll get rid

mcollina commented on lib/auth.js:
```diff
@@ -0,0 +1,74 @@
+
+  [kMakeProtectedResolver] (authDirectiveAST, resolverFn) {
+    return async (parent, args, context, info) => {
+      const result = await this[kApplyPolicy](authDirectiveAST, parent, args, context, info)
```
why somebody would return an Error instead of throwing?

jonnydgreen replied:
I wanted to replicate what is possible within a `graphql-js` resolver - something I learnt yesterday was that GraphQL execute handles errors returned in resolvers as well as when they are thrown, e.g.:

```js
const resolvers = {
  Query: {
    add: async (_, obj) => {
      return new Error('kaboom')
    }
  }
}
```

However, I agree with you that just throwing makes sense and keeps things cleaner so I'll remove this handling! :)

mcollina replied:
Oh I didn't know this. Then +1 in doing this as well. Maybe add a comment?

jonnydgreen replied:
Yeah I didn't know about it either! Cool, yep no problem - I'll add a comment! :)

mcollina reviewed (changes requested):
Really good progress so far! It needs docs but this is phenomenal progress!

jonnydgreen commented:
Thanks!! :) Yep, starting the docs and examples today!

jonnydgreen commented:
Latest benchmarks on my local machine (gateway is ~50% quicker!):

```txt
==============================
= Normal Mode | Without Auth =
==============================
┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
│ Latency │ 4 ms │ 5 ms │ 9 ms  │ 13 ms │ 5.17 ms │ 2.27 ms │ 74 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%    │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 10479   │ 10479   │ 18447  │ 18687   │ 17582.8 │ 2388.24 │ 10475   │
├───────────┼─────────┼─────────┼────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 5.51 MB │ 5.51 MB │ 9.7 MB │ 9.83 MB │ 9.25 MB │ 1.26 MB │ 5.51 MB │
└───────────┴─────────┴─────────┴────────┴─────────┴─────────┴─────────┴─────────┘
Req/Bytes counts sampled once per second.
176k requests in 10.03s, 92.5 MB read

===========================
= Normal Mode | With Auth =
===========================
┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
│ Latency │ 5 ms │ 5 ms │ 10 ms │ 15 ms │ 5.68 ms │ 2.16 ms │ 66 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
┌───────────┬────────┬────────┬─────────┬─────────┬──────────┬─────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min    │
├───────────┼────────┼────────┼─────────┼─────────┼──────────┼─────────┼────────┤
│ Req/Sec   │ 9703   │ 9703   │ 17007   │ 17455   │ 16327.64 │ 2109.39 │ 9699   │
├───────────┼────────┼────────┼─────────┼─────────┼──────────┼─────────┼────────┤
│ Bytes/Sec │ 5.1 MB │ 5.1 MB │ 8.95 MB │ 9.18 MB │ 8.59 MB  │ 1.11 MB │ 5.1 MB │
└───────────┴────────┴────────┴─────────┴─────────┴──────────┴─────────┴────────┘
Req/Bytes counts sampled once per second.
180k requests in 11.03s, 94.5 MB read

===============================
= Gateway Mode | Without Auth =
===============================
┌─────────┬───────┬───────┬───────┬───────┬──────────┬──────────┬────────┐
│ Stat    │ 2.5%  │ 50%   │ 97.5% │ 99%   │ Avg      │ Stdev    │ Max    │
├─────────┼───────┼───────┼───────┼───────┼──────────┼──────────┼────────┤
│ Latency │ 29 ms │ 32 ms │ 68 ms │ 94 ms │ 34.77 ms │ 11.83 ms │ 191 ms │
└─────────┴───────┴───────┴───────┴───────┴──────────┴──────────┴────────┘
┌───────────┬────────┬────────┬─────────┬─────────┬────────┬────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%     │ 97.5%   │ Avg    │ Stdev  │ Min    │
├───────────┼────────┼────────┼─────────┼─────────┼────────┼────────┼────────┤
│ Req/Sec   │ 1250   │ 1250   │ 3081    │ 3161    │ 2833.9 │ 567.6  │ 1250   │
├───────────┼────────┼────────┼─────────┼─────────┼────────┼────────┼────────┤
│ Bytes/Sec │ 438 kB │ 438 kB │ 1.08 MB │ 1.11 MB │ 992 kB │ 199 kB │ 438 kB │
└───────────┴────────┴────────┴─────────┴─────────┴────────┴────────┴────────┘
Req/Bytes counts sampled once per second.
28k requests in 10.03s, 9.92 MB read

============================
= Gateway Mode | With Auth =
============================

┌─────────┬───────┬───────┬───────┬───────┬──────────┬──────────┬────────┐
│ Stat    │ 2.5%  │ 50%   │ 97.5% │ 99%   │ Avg      │ Stdev    │ Max    │
├─────────┼───────┼───────┼───────┼───────┼──────────┼──────────┼────────┤
│ Latency │ 29 ms │ 33 ms │ 66 ms │ 96 ms │ 35.83 ms │ 12.74 ms │ 213 ms │
└─────────┴───────┴───────┴───────┴───────┴──────────┴──────────┴────────┘
┌───────────┬────────┬────────┬─────────┬─────────┬────────┬────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%     │ 97.5%   │ Avg    │ Stdev  │ Min    │
├───────────┼────────┼────────┼─────────┼─────────┼────────┼────────┼────────┤
│ Req/Sec   │ 1210   │ 1210   │ 2969    │ 3089    │ 2750.7 │ 548.59 │ 1210   │
├───────────┼────────┼────────┼─────────┼─────────┼────────┼────────┼────────┤
│ Bytes/Sec │ 424 kB │ 424 kB │ 1.04 MB │ 1.08 MB │ 962 kB │ 192 kB │ 424 kB │
└───────────┴────────┴────────┴─────────┴─────────┴────────┴────────┴────────┘
Req/Bytes counts sampled once per second.
28k requests in 10.03s, 9.63 MB read
```

mcollina commented:
I do not see those changes here! Remember to remove the workers option as well.

jonnydgreen commented:
Docs and examples now done! For the benchmarking performance, enabling JIT had a significant impact (~50% faster) - thanks for releasing Mercurius with my JIT hooks fix to enable that! :) Other options I tried didn't have any noticeable impact, but I will keep experimenting!

mcollina reviewed (commented):
Good work! I might have missed it as there is a lot of content, but I do not see an example that shows how to fulfill a slight more complex permission use case.

Let's imagine we have Alice that is part of organization Acme. Bob is not part of Acme. Both are users of the platform. We want to restrict Bob to perform any operations on Acme.

jonnydgreen commented:
Sure thing, yeah I can add an advanced use case like you suggested - no problem!

jonnydgreen commented:
Added your suggestion with examples, tests and docs as well as another advanced example where one registers multiple auth plugins. I've also added the application hook for refreshing the auth overrides and tidied up the code a bit more!

mcollina commented:
could we make the `authDirective` option just a string?

jonnydgreen commented:
Yeah no problem, I guess that makes some of the logic a bit cleaner and we don't lose anything from a functionality point of view - I'll get on that first thing tomorrow!

jonnydgreen commented:
Sorted, took me a bit longer than I thought as I'd been using GraphQLDirective instances everywhere! Should be sorted now from a docs,tests,examples point of view as well

mcollina commented on bench/normal-setup.js:
```diff
@@ -0,0 +1,76 @@
+}`
+
+const schema = `
+  ${authDirective}
```
Why are we duplicating this in two places? Can we avoid it?

jonnydgreen replied:
when you say duplicating, are you referring to this code being used in both `normal-with-auth.js and normal-wihout-auth.js`? Or the fact that it's very similar to code in the `examples` (which I could benchmark against instead to be fair)?

mcollina replied:
No. I mean the `auth` is being passed in inside the schema and as an option of the module. We might be able to skip it entirely? Maybe we can just pass in the name of the directive to fetch from the parsed schema?

jonnydgreen replied:
Yeah that makes sense as we only use the name anyway and directive names are unique within the schema - it also has the benefit of preventing users from defining multiple directives within the definition they are providing! I'll get on that asap!

mcollina commented:
(I know I'm picky, thanks for bearing with me).

jonnydgreen commented:
Haha, no worries! I'm exactly the same :)

mcollina reviewed (approved):
lgtm, this is massive work.

Are you planning to do a blog post about this?

mcollina commented:
go for it! Add me as an owner afterwards (https://www.npmjs.com/~matteo.collina)

jonnydgreen commented:
Cool, thanks! (Sorry for the delay, I missed the notification from GH for some reason)
